### PR TITLE
updated the host url

### PR DESCRIPTION
**Pull request by Asheshp23**

### CODE REVIEW

File: Sources/SimpleSalesforceSDK/Services/WebService.swift [hunk 1]
```diff
@@ -96,8 +100,12 @@ class WebService {
       let jsonData = try? JSONSerialization.data(
         withJSONObject: fieldUpdates, options: .prettyPrinted)
 
-      let url = "\(config.host)/data/v54.0/sobjects/\(objectName)/\(id)"
-      let fetchUrl = URL(string: "\(url)")!
+      let url = "\(config.host)services/data/v54.0/sobjects/\(objectName)/\(id)"
+     guard let fetchUrl = URL(string: "\(url)")
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Services/WebService.swift]: fix indentation

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 2]
```diff
@@ -91,7 +91,11 @@ class SFAuth {
     "&client_id=\(config.clientId)" +
     "&refresh_token=\(refreshToken)"
     
-    guard let url = URL(string: "\(config.host)/oauth2/token") else { return }
+    guard let url = URL(string: "\(config.host)services/oauth2/token")
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift]: What if the host supplied does not end with `/`?

I found this SO answer https://stackoverflow.com/questions/17419460/should-i-save-a-path-with-or-without-a-trailing-slash-at-the-end-whats-the-con

If we expect something lets make sure we either validate it at the time of configure call or actually modify the incoming url and add the trailing slash if the sdk user missed it

File: Sources/SimpleSalesforceSDK/Models/Errors.swift [hunk 3]
```diff
@@ -7,6 +7,7 @@ public enum SSSDKError: Error {
   case authNoRefreshTokenError
   case authRefreshFailedError
   case authIntrospectFailedError
+  case invalidUrlError
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/Errors.swift]: maybe this should be 
```suggestion
  case invalidUrlError(url: String)
```

That way you can include the url in the error for the sdk user to inspect?

File: Sources/SimpleSalesforceSDK/Models/Errors.swift [hunk 4]
```diff
@@ -23,6 +24,8 @@ extension SSSDKError: CustomStringConvertible {
         return "Refreshing access token failed, try login again."
       case .authIntrospectFailedError:
         return "Introspecting access token failed, try login again."
+      case .invalidUrlError:
+        return "The provided URL is invalid"
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/Errors.swift]:
```suggestion
      case .invalidUrlError(let url):
        return "The provided URL (\(url)) is invalid"
```

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  var bearerToken: String?
+  var contentType: String
+}
+
+public struct URLRequestBuilder {
+  ///  Creates URL for updateData api
+  /// - Parameters:
+  ///     - requestConfig: The URLRequestConfig instance’s configuration.
+  /// - Returns: returns URLRequest for requested URLRequestConfiguration.
+  public static func request(requestConfig: URLRequestConfig) -> URLRequest {
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]:
```suggestion
  public static func request(with config: URLRequestConfig) -> URLRequest {
```

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import Foundation
+
+public struct URLRequestConfig {
+  var url: URL
+  var params: Data?
+  var httpMethod: String
+  var bearerToken: String?
+  var contentType: String
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]: Maybe have defaults for most common cases
```suggestion
  var contentType: String = "application/json"
```

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import Foundation
+
+public struct URLRequestConfig {
+  var url: URL
+  var params: Data?
+  var httpMethod: String
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]: Default to `get`?
```suggestion
  var httpMethod: String = HTTP_METHOD_GET
```

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import Foundation
+
+public struct URLRequestConfig {
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]: maybe add your http verbs as constants?

```
public static let HTTP_METHOD_GET = "GET"
```

avoids typos

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    if let bearerToken = requestConfig.bearerToken {
+      request.setValue(bearerToken, forHTTPHeaderField: "Authorization")
+    }
+    if let params =  requestConfig.params {
+    request.httpBody = params
+    }
+
+    return request
+  }
+
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]: remove empty line

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    var request = URLRequest(url: requestConfig.url)
+    request.httpMethod = requestConfig.httpMethod
+    request.setValue(requestConfig.contentType,
+                     forHTTPHeaderField: "Content-Type")
+
+    if let bearerToken = requestConfig.bearerToken {
+      request.setValue(bearerToken, forHTTPHeaderField: "Authorization")
+    }
+    if let params =  requestConfig.params {
+    request.httpBody = params
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]: indentation

**Comment by Asheshp23** [comment]
- Added enum for HTTPMethod and Content Type.
- Fixed Indentation.

File: Sources/SimpleSalesforceSDK/Models/SFAuth.swift [hunk 11]
```diff
@@ -130,18 +133,21 @@ class SFAuth {
...
-    request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
-    request.httpBody = params.data(using: .utf8)
-    
+
+    let request = URLRequestBuilder
+      .request(with:
+                URLRequestConfig(url: url,
+                                 params: params.data(using: .utf8),
+                                 httpMethod: .post,
+                                 contentType: .urlEncoded))
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/SFAuth.swift:149]: 😍

File: Sources/SimpleSalesforceSDK/Models/URLBuilder.swift [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  /// - Throws: `SSSDKError.invalidUrlError` if the provided host url is invalid
+  /// - Returns: returns URL for refresh access token api.
+  public static func refreshTokenURL(urlString: String) throws -> URL {
+    let host = verifyHost(host: urlString)
+    guard let refreshToneUrl = URL(string: "\(host)services/oauth2/token")
+    else {
+      throw SSSDKError.invalidUrlError(url: host)
+    }
+    return refreshToneUrl
+
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLBuilder.swift]:
```suggestion
    guard let refreshTokenUrl = URL(string: "\(host)services/oauth2/token")
    else {
      throw SSSDKError.invalidUrlError(url: host)
    }
    return refreshTokenUrl
```

File: Sources/SimpleSalesforceSDK/Models/URLBuilder.swift [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  }
+  ///  Creates URL for fetchData api
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLBuilder.swift:61]: be consistent with whether to add spaces after methods or not, PLEASE!

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    case post = "POST"
+    case put = "PUT"
+    case delete = "DELETE"
+    case connect = "CONNECT"
+    case options = "OPTIONS"
+    case trace = "TRACE"
+    case patch = "PATCH"
+  }
+  
+  enum CONTENTTYPE: String {
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]: Why all caps?
```suggestion
  enum RequestContentType: String {
```

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 15]
```diff
@@ -0,0 +1,48 @@
+import Foundation
+
+public struct URLRequestConfig {
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift:3]: Don't think this should have `URL` in the name...

```suggestion
public struct RequestConfig {
```

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  }
+  
+  enum CONTENTTYPE: String {
+    case json = "application/json"
+    case urlEncoded = "application/x-www-form-urlencoded"
+  }
+}
+
+public struct URLRequestBuilder {
+  ///  Creates URL for updateData api
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]: Wrong comment...
```suggestion
public struct URLRequestBuilder {
  ///  Builds URLRequests based on provided config
```

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+  /// - Parameters:
+  ///     - requestConfig: The URLRequestConfig instance’s configuration.
+  /// - Returns: returns URLRequest for requested URLRequestConfiguration.
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]:
```suggestion
  /// - Returns: returns URLRequest for requested URLRequestConfig.
```

File: Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+  enum CONTENTTYPE: String {
+    case json = "application/json"
+    case urlEncoded = "application/x-www-form-urlencoded"
+  }
+}
+
+public struct URLRequestBuilder {
+  ///  Creates URL for updateData api
+  /// - Parameters:
+  ///     - requestConfig: The URLRequestConfig instance’s configuration.
```
Comment (ipranay) [inline Sources/SimpleSalesforceSDK/Models/URLRequestBuilder.swift]:
```suggestion
  ///     - config: Configuration for the request
```

**Review by ipranay — APPROVED** [review]
Few minor comments, 👍 once they're fixed

**Pull request merged**